### PR TITLE
TW-1519: Automatically synchronize pin unpin

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket

- #1519

### NEED MERGE:

- https://github.com/linagora/matrix-dart-sdk/pull/51

#### WEB

- **CHAT** 

https://github.com/linagora/twake-on-matrix/assets/99852347/a2348edc-eb27-4691-b2c5-35a74547af78

https://github.com/linagora/twake-on-matrix/assets/99852347/86f7b5b5-7b2e-4b0e-be5b-7fa872a400cd

- **PIN SCREEN**

https://github.com/linagora/twake-on-matrix/assets/99852347/9b1ff3dd-1a12-46ab-abde-b498e7b3effa

https://github.com/linagora/twake-on-matrix/assets/99852347/e46c9a91-75f4-4d35-b652-dfdfa872b917

#### MOBILE

https://github.com/linagora/twake-on-matrix/assets/99852347/d562f2a3-b1f4-44b3-bd91-c4a4e2419c0a

https://github.com/linagora/twake-on-matrix/assets/99852347/c7628779-fbe5-43be-a606-f060c2fabca5

### CODE REVIEW

File: lib/pages/chat/chat.dart [hunk 1]
```diff
@@ -1705,6 +1706,40 @@ class ChatController extends State<Chat>
...
   }
 
+  void _listenRoomUpdateEvent() {
+    if (room == null) return;
+    client.onEvent.stream.listen((eventUpdate) {
+      Logs().d(
+        'Chat::_listenRoomUpdateEvent():: Event Update Content ${eventUpdate.content}',
+      );
+      if (eventUpdate.isPinnedEventsHasChanged) {
+        WidgetsBinding.instance.addPostFrameCallback((_) async {
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1716]: remove async

File: lib/pages/chat/chat.dart [hunk 2]
```diff
@@ -1705,6 +1706,40 @@ class ChatController extends State<Chat>
...
+      }
+    });
+  }
+
+  void _handlePinnedMessageCallBack({
+    required bool isInitial,
+    required bool isUnpin,
+    String? eventId,
+  }) {
+    if (room == null) return;
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1730]: check `roomId` also

File: lib/pages/chat/chat_pinned_events/pinned_events_controller.dart [hunk 3]
```diff
@@ -97,6 +97,7 @@ class PinnedEventsController {
     );
     if (index < 0) {
       currentPinnedEventNotifier.value = pinnedEvents.first;
+      return 0;
```
Comment (sherlockvn) [inline lib/pages/chat/chat_pinned_events/pinned_events_controller.dart:100]: create a method `handleWhenNotFoundPinnedMessage`

File: lib/pages/chat/chat_pinned_events/pinned_messages.dart [hunk 4]
```diff
@@ -359,11 +362,56 @@ class PinnedMessagesController extends State<PinnedMessages>
...
   }
 
+  void _listenRoomUpdateEvent() {
+    if (room == null) return;
+    client.onEvent.stream.listen((eventUpdate) {
+      Logs().d(
+        'PinnedMessages::_listenRoomUpdateEvent():: Event Update Content ${eventUpdate.content}',
+      );
+      if (eventUpdate.isPinnedEventsHasChanged) {
+        WidgetsBinding.instance.addPostFrameCallback((_) async {
```
Comment (sherlockvn) [inline lib/pages/chat/chat_pinned_events/pinned_messages.dart:372]: remove async

File: lib/pages/chat/chat_pinned_events/pinned_messages.dart [hunk 5]
```diff
@@ -359,11 +362,56 @@ class PinnedMessagesController extends State<PinnedMessages>
...
+      }
+
+      if (eventsNotifier.value.isNotEmpty && result.isEmpty) {
+        eventsNotifier.value = result;
+        Navigator.of(context).pop();
+      }
+    } on MatrixException catch (exception) {
+      Logs().e(
+        'PinnedMessages::_handlePinnedMessageCallBack():: ErrorCode ${exception.errcode}: ${exception.errorMessage}',
+      );
```
Comment (sherlockvn) [inline lib/pages/chat/chat_pinned_events/pinned_messages.dart:405]: should we show something to the user ?

File: lib/presentation/extensions/event_update_extension.dart [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+  void updatePinnedMessage({
+    required PinnedMessageCallBack pinnedMessageCallBack,
+  }) {
+    if (!isPinnedEventsHasChanged) return;
+    final content = this.content['content'];
+    final prevContent = this.content['unsigned']['prev_content'];
+    if (content != null && prevContent != null) {
+      final pinnedList = content['pinned'] as List<dynamic>;
+      final prevPinnedList = prevContent['pinned'] as List<dynamic>;
```
Comment (sherlockvn) [inline lib/presentation/extensions/event_update_extension.dart:31]: check before cast

**Review by Te-Z — APPROVED** [review]

**Review by hieutbui — APPROVED** [review]

File: lib/presentation/extensions/event_update_extension.dart [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      return id;
+    }
+    return null;
+  }
+
+  bool get isPinnedEventsHasChanged {
+    return content['type'] == EventTypes.RoomPinnedEvents;
+  }
+
+  void updatePinnedMessage({
```
Comment (hoangdat) [inline lib/presentation/extensions/event_update_extension.dart:22]: separate into small method

File: lib/presentation/extensions/event_update_extension.dart [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import 'package:matrix/matrix.dart';
+
+typedef PinnedMessageCallBack = void Function({
```
Comment (hoangdat) [inline lib/presentation/extensions/event_update_extension.dart:3]: `onPinnedMessageUpdated`

File: lib/presentation/extensions/event_update_extension.dart [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  bool get isPinnedEventsHasChanged {
+    return content['type'] == EventTypes.RoomPinnedEvents;
+  }
+
+  void updatePinnedMessage({
+    required PinnedMessageCallBack pinnedMessageCallBack,
+  }) {
+    if (!isPinnedEventsHasChanged) return;
+    final content = this.content['content'];
+    final prevContent = this.content['unsigned']['prev_content'];
```
Comment (sherlockvn) [inline lib/presentation/extensions/event_update_extension.dart:27]: this.content['unsigned'] may be null

**Review by hoangdat — APPROVED** [review]

**Pull request merged**